### PR TITLE
Improve the developer experience through better Invoke commands

### LINKED ISSUE #92: Improve the Developer Experience
A couple of improvements we can bring to `invoke <command>` for a better DX:
- [x] Ensure that initial requirements (Invoke & Poetry) can be brought without creating a Poetry _.venv_ first.
- [x] #255
- [x] Commands depending on the local virtual environment should be going through Poetry's context. (e.g. `inv make-release` fails to correctly detect the current version, where `poetry run inv make-release` works)
- [x] #262
- [x] #264
- [ ] #261
- [ ] #263 
- [x] #287 
- [x] #299
- [x] #300
- [x] #337 
- [ ] #267

📝 At the end, we should have a clear development guide teaching you in what order these commands should be run during the development journey.

### LINKED ISSUE #255: Improve Invoke commands
According to [Notion](https://www.notion.so/robocorptechnologies/DX-improvements-519c97b7bf7c4bf097b2ca0885a4e51b?pvs=4)

### PR DESCRIPTION
**Pull request by cmin764**
Closes #255 

- More robust way of defining the initial dependencies which will setup the package environment. (plain `inv <cmd>` working in all the packages)
- Improved Invoke commands according to [Notion](https://www.notion.so/robocorptechnologies/DX-improvements-519c97b7bf7c4bf097b2ca0885a4e51b?pvs=4).
- Fixes the `meta` package way of importing and using `devutils`.

### CODE REVIEW

File: devutils/pyproject.toml
Comment (fabioz) [inline devutils/pyproject.toml:24]: I'm not sure I follow, what do we gain by removing things from pyproject.toml?

File: excel/tasks.py
Comment (fabioz) [inline excel/tasks.py]: I'm not sure this is an improvement... it works because the `invoke_utils` doesn't import anything outside of that file at this point, but it'll break when that happens.

File: devutils/pyproject.toml
Comment (cmin764) [reply devutils/pyproject.toml:24]: Was thinking that this might take away from the confusion of who is responsible for providing you with the Invoke tool. Is it yourself initially or the Poetry-based venv after a `poetry install`? (this way you'd be sure that you should get Invoke from the _requirements.txt_ only and that is something living outside your package _.venv_)

Also it helps with pinning & maintaining a compatible version in a single place, which is _requirements.txt_. But if you'll need to use Invoke from Poetry's context (e.g.: `inv <task>` -> `poetry(..., "inv", ...)`), then yes, it might make sense to keep that dep here if the initial executable isn't reachable anymore.

Will add it back if it doesn't produce any harm or confusion, as I don't have a strong opinion either. And now this stems another question:
> What do we gain by adding unnecessary dependencies in **devtools**' _pyproject.toml_?

File: excel/tasks.py
Comment (cmin764) [reply excel/tasks.py]: Correct, I missed that!

Just that changing the `sys.path` at runtime (and such change has to be maintained in a dozen of different _tasks.py_ files) didn't look like a better idea than this newer suggestion I added. (and improved post-review)

----

I added a few more fixes to strengthen the solution and make it cleaner, so that:
- Any _tasks.py_ file doesn't even need to know about the files/modules we write in _devutils_, it is sufficient now to refer to: `"devutils" / "src" / "devutils" / "__init__.py"`
- Any "publicly" exposed function at package init level is directly usable by any importer: `devutils.build_common_tasks()`
- The _invoke_utils.py_ file is the only one in charge now for managing `sys.path` in a single place, with a path relative to itself, and now we can even do imports like: `from devutils.rcc import build_subprocess_kwargs`

What do you think now, isn't [this](https://github.com/robocorp/robocorp/blob/chore/dx-improvements/excel/tasks.py#L20-L23) more cleaner, robust and easier to maintain now in all those _tasks.py_ files we have in every package?

File: devutils/src/devutils/invoke_utils.py
Comment (bogdancondurache) [inline devutils/src/devutils/invoke_utils.py]: one idea, might be premature optimisation at this point, would be to have this part in the __init__.py file and then you would be able to use `devutils` imports from any file, not just this one. Of course, at the moment it's just this file, but in the future if somebody adds a new file and the imports won't work he/she might be very confused as to why :)

File: devutils/pyproject.toml
Comment (fabioz) [reply devutils/pyproject.toml:24]: > What do we gain by adding unnecessary dependencies in devtools' pyproject.toml?

Nothing, but as you mentioned before:

> But if you'll need to use Invoke from Poetry's context (e.g.: inv <task> -> poetry(..., "inv", ...)), then yes, it might make sense to keep that dep here if the initial executable isn't reachable anymore.

So, we do get something here (although our automations using `inv` make it so that it doesn't need the poetry env active, sometimes I do expect `inv` or `poetry` to be in the env if I'm doing things that are not initially in our automations).

File: excel/tasks.py
Comment (bogdancondurache) [inline excel/tasks.py]:
```suggestion
# environment only, we won't have it available with Invoke commands running outside
# such environment.
```

File: devutils/src/devutils/invoke_utils.py
Comment (cmin764) [reply devutils/src/devutils/invoke_utils.py]: Yes and thank you for the notice!

I'll revert all of these anyway as it was proved that they bring more complexities than clarity. Will ask for a proper review once I mark this as non-draft anymore :).

File: excel/tasks.py
Comment (cmin764) [reply excel/tasks.py]: There's a rule in PyCharm where such indentation separates a contiguous comment block from the other comments.

Highlight can be observed with TODO and FIXME:
Screenshot 2024-02-08 at 4 05 50 PM
Screenshot 2024-02-08 at 4 05 57 PM

File: excel/tasks.py
Comment (cmin764) [reply excel/tasks.py]: Reverted as it was proved that `devutils` isn't expected to be a cohesive Python package and my change didn't add value.

File: excel/tasks.py
Comment (bogdancondurache) [reply excel/tasks.py]: interesting, had no idea this is a normal format

File: devutils/src/devutils/invoke_utils.py
Comment (fabioz) [inline devutils/src/devutils/invoke_utils.py]: Please don't `resolve()` here as it'll resolve symlinks which is not what we want.

Use `absolute()` instead.

File: meta/tasks.py
Comment (fabioz) [inline meta/tasks.py]: Don't `resolve()`. Use `absolute()` instead.

Note: you almost never want to `resolve()` unless you need a unique reference to some file and want to create some sort of cache.

The reason for this is because when symlinks are resolved it'll change the view of the working dir.

For instance, if you are in a subst folder such as:

`x:`

which points to `d:\my_x_subst`

And the .pyc was generated with the file pointing to `x:\my_file.py`, if you resolve() it'll actually become `d:\my_y_subst\my_file.py` and you'd break any view which was tailored by the user for the working dir being used (at best -- probably the case here -- you'd just start getting the .pycs from a different `__file__`, but at worst you would end up having imports that no longer resolve because that different view of the structure is no longer true anymore).

File: devutils/src/devutils/invoke_utils.py
Comment (fabioz) [inline devutils/src/devutils/invoke_utils.py:206]: Where did `devinstall` go? It should still be there...

Comment (OvidiuCode) [reply devutils/src/devutils/invoke_utils.py:206]: We delete it in favor of --local argument on install. But if you think we need a standalone command we can put it back.

Comment (fabioz) [reply devutils/src/devutils/invoke_utils.py:206]: I'd rather have a standalone command -- and then you won't need to update all the github actions to use the new version too ;)

Comment (OvidiuCode) [reply devutils/src/devutils/invoke_utils.py:206]: ok, I'll add it back

File: devutils/src/devutils/invoke_utils.py
Comment (cmin764) [inline devutils/src/devutils/invoke_utils.py]:
```suggestion
            prefix.append("-vv")
```
as one level of verbosity is not enough here. (doesn't spit too much)

File: devutils/src/devutils/invoke_utils.py
Comment (cmin764) [inline devutils/src/devutils/invoke_utils.py]:
```suggestion
    def install(ctx, local: Optional[str] = None, update: bool = False, verbose: bool = False):
```

File: devutils/src/devutils/invoke_utils.py
Comment (cmin764) [reply devutils/src/devutils/invoke_utils.py:206]: Good point!

File: devutils/src/devutils/invoke_utils.py
Comment (cmin764) [inline devutils/src/devutils/invoke_utils.py]:
```suggestion
    def mark_as_develop_mode(projects: Optional[list[str]] = None, all_packages: bool = False):
```
if we still want to stay consistent with typing. Also is not a good idea to shadow the builtin `all`.

File: devutils/src/devutils/invoke_utils.py
Comment (cmin764) [inline devutils/src/devutils/invoke_utils.py]:
```suggestion
    def test(ctx, test: Optional[str] = None):
```

File: devutils/src/devutils/invoke_utils.py
Comment (cmin764) [inline devutils/src/devutils/invoke_utils.py]: I would also print the `unreleased_changes` as well, so I would have a confirmation of what I'm releasing next, as the file that gets changed is always the same and doesn't say too much.

File: excel/tasks.py
Comment (cmin764) [inline excel/tasks.py:1]: If we like this refactor (which adds a little bit more clarity imho), then we can copy-paste it to the other _tasks.py_ files as well. (doesn't change behaviour, just structure)

File: devutils/src/devutils/invoke_utils.py
Comment (cmin764) [inline devutils/src/devutils/invoke_utils.py]: This `all` is a reserved Python keyword (the builtin function), I would recommend replacing it with something like `all_packages` (or similar) to avoid name shadowing.

File: devutils/src/devutils/invoke_utils.py
Comment (cmin764) [inline devutils/src/devutils/invoke_utils.py]: 🐛 Bug spotted:
```suggestion
                        if all or key in projects:
                            dependencies[key] = dict(path=f"../{name}/", develop=True)
```

- `name`: `workitems`
- `key`: `robocorp-workitems`
- `projects`: [`robocorp-workitems`]

So you should check the `key` belonging to the list, not the `name`, which always fails to be found.

**Review by cmin764 — COMMENTED** [review]
LGTM overall and thank you @OvidiuCode for your hard work!

⚠️ Still, I want to request some changes before merging as pointed in the comments below this one.

----

Once this gets merged (or about to be merged), please check what workflows need to be changed/updated to match the new state of the upcoming Invoke commands. Something like:

```
    - name: Install poetry
       run: pipx install poetry
    ....
    - name: Install invoke/tomlkit
       run: pip install invoke tomlkit
```
->
```
    - name: Install development dependencies
       run: pipx install -Ur requirements.txt  # from devutils
```

`python -m invoke <cmd>` -> `inv <cmd>`
if works and simpler to use

```
    - name: Test
      env:
        GITHUB_ACTIONS_MATRIX_NAME: ${{ matrix.name }}
        CI_CREDENTIALS: ${{ secrets.CI_CREDENTIALS }}
        CI_ENDPOINT: ${{ secrets.CI_ENDPOINT }}
      run: python -m invoke test
    - name: '`inv lint`/`inv typecheck`, potentially fixed with `inv pretty`'
      if: always()
      run: '

        python -m invoke lint

        python -m invoke typecheck

        '
```
->
```
    - name: Lint and test
      env:
        GITHUB_ACTIONS_MATRIX_NAME: ${{ matrix.name }}
        CI_CREDENTIALS: ${{ secrets.CI_CREDENTIALS }}
        CI_ENDPOINT: ${{ secrets.CI_ENDPOINT }}
      run: inv check-all
```
since we're checking in one go now: linting, typing, tests and docs. (and we were checking linting & typing anyways for the entire matrix before)


```
    - name: Check tag version
      run: poetry run invoke check-tag-version
```
->
```
    - name: Check tag version
      run: inv check-tag-version
```

```
    - name: Build with poetry
      run: poetry build

    - name: Upload to PyPI
      run: |
        poetry config pypi-token.pypi  ${{ secrets.PYPI_TOKEN_ROBOCORP_WORKITEMS }}
        poetry publish
```
->
```
    - name: Build package and upload to PyPI
      run: |
        inv build
        inv publish --token ${{ secrets.PYPI_TOKEN_ROBOCORP_WORKITEMS }}
```
once we add such additional `pubish` Invoke command.

File: devutils/src/devutils/invoke_utils.py
Comment (fabioz) [inline devutils/src/devutils/invoke_utils.py]: `print(print(...))` seems odd here...

File: devutils/src/devutils/invoke_utils.py
Comment (fabioz) [inline devutils/src/devutils/invoke_utils.py:497]: What's the expected format of the changelog that this will act upon?

The format I was usually working on so far was:

```
# Changelog

## Unreleased

- xxx
- yyy

## 0.0.20 - 2024-01-31

- zzz
```

But then I'd expect this code here to just change the 'Unreleased' to the version here, but this isn't what the code is doing...

File: devutils/src/devutils/invoke_utils.py
Comment (cmin764) [reply devutils/src/devutils/invoke_utils.py]: Yes, it prints an additional `None` because of that, just spotted it during testing.

File: devutils/src/devutils/invoke_utils.py
Comment (fabioz) [reply devutils/src/devutils/invoke_utils.py:497]: I think it may be nicer to change the code to deal with that structure and raise an error if `## Unreleased` is not found right after `## Changelog` complaining that the changelog is not up to date.

File: devutils/src/devutils/invoke_utils.py
Comment (cmin764) [inline devutils/src/devutils/invoke_utils.py:515]:
```suggestion
    def set_version(ctx, version: str):
```

Would be awesome if you can parse the version and check if it's numeric, as I'm always accidentally running: `inv set-version -h` to see some usage and guess what, I end up with this version:
```
__version__ = "-h"
```

You can use this [pattern](https://github.com/robocorp/rpaframework/blob/master/packages/core/src/RPA/core/certificates.py#L45): `packaging.version.parse()` to parse and check the atoms if are integers. (or at least 3 in number, as we're following semver: `major.minor.hotfix`)

File: excel/tasks.py
Comment (fabioz) [reply excel/tasks.py:1]: I agree that having a variable to don't duplicate the `Path(__file__).absolute().parent` is a bit better -- but I'm also not bothered enough by how it was previously to manually go on to change all the files ;)

File: meta/tasks.py
Comment (cmin764) [reply meta/tasks.py]: Maybe that's why I was encountering some issues during Windows development which was accessing files on a mounted drive enabling access to files on my Mac, thing which was messing a little with my drive letters. Although not sure if `absolute()` will still break it.

Was thinking to use `resolve()` just in case we end up with `..` (parent) or `~` (home) in the path somehow.

File: devutils/src/devutils/invoke_utils.py
Comment (cmin764) [reply devutils/src/devutils/invoke_utils.py:497]: I removed long-time ago the `## Unreleased` from every CHANGELOG because of inconsistencies and I believed that we don't want to use that.

But if it makes more sense to keep that sentinel header, then also fine by me, thus we can re-introduce it everywhere and adapt the code to work on it.

File: devutils/src/devutils/invoke_utils.py
Comment (OvidiuCode) [reply devutils/src/devutils/invoke_utils.py]: projects is actually the list we pass on which is `workitems, tasks` and so on. I think it would be annoying to write the `robocorp-` suffix for all of our packages we want to install in devmode.

File: devutils/src/devutils/invoke_utils.py
Comment (OvidiuCode) [reply devutils/src/devutils/invoke_utils.py]: nice catch!

File: devutils/src/devutils/invoke_utils.py
Comment (OvidiuCode) [reply devutils/src/devutils/invoke_utils.py:497]: I don't have a strong opinion here either. We can add back the `Unreleased` header or just add the comments above the latest version. But we need to decide and be consistent.

File: devutils/src/devutils/invoke_utils.py
Comment (OvidiuCode) [reply devutils/src/devutils/invoke_utils.py]: I don't think that is very important either because you can check it with a git diff and if we decide to have the `## Unrealeased` header then we don't need that piece of logic anymore.

File: meta/tasks.py
Comment (fabioz) [reply meta/tasks.py]: > Maybe that's why I was encountering some issues during Windows development which was accessing files on a mounted drive enabling access to files on my Mac, thing which was messing a little with my drive letters. Although not sure if absolute() will still break it.

Definitely (I do have many scars on that -- `resolve()` is very seldomly needed and is usually the source of bugs). `absolute()` will do the `..` -- for `~` you need `expanduser()` (but in this case we can leave `~` if it ever gets there as there's no issue in that, the `absolute()` is only needed because if `.` was there then we could have a problem).

File: devutils/src/devutils/invoke_utils.py
Comment (fabioz) [reply devutils/src/devutils/invoke_utils.py:497]: The good thing about `## Unreleased` -- maybe accepting any variation with any casing -- is that it gives a clear point where that'd be needed and it'd be a good point of failure.

With the current code it'd need to do more analysis to decide if the format is valid for applying the header (i.e.: it'll happily apply the header with no changes or when done a second time for the same version, having the `unreleased` would make the implementation a bit easier -- although the implementation can also be improved to handle those).

File: meta/tasks.py
Comment (cmin764) [reply meta/tasks.py]: Looks like a strong opinion based on data, so +1 on this!

File: devutils/src/devutils/invoke_utils.py
Comment (cmin764) [reply devutils/src/devutils/invoke_utils.py]: Indeed, for some reason, during testing, I thought it doesn't do the job if you pass `-l workitems` for e.g., so I fooled myself into thinking that's a bug when I tried with `-l robocorp-workitems`.

To be on the safe side, you can apply a `projects = [package.replace("robocorp-", "").strip() for package in (local or "").split(",")]` so no matter if somebody passes it with `robocorp-` prefix or not, it would still work in the normalized way.

File: devutils/src/devutils/invoke_utils.py
Comment (cmin764) [inline devutils/src/devutils/invoke_utils.py:465]: Post-merge, you can test this out by making a release that we really need for `excel` (which I missed to do), which will produce this tag: `robocorp-excel-0.4.3`

As currently `inv check-tag-version` in `excel`:
```
Version does not match (robocorp-excel: 0.4.3 != repo tag: 0.4.2).
Tags:['robocorp-excel-0.3.0', 'robocorp-excel-0.4.0', 'robocorp-excel-0.4.1', 'robocorp-excel-0.4.2']
(exit(1))
```

So this is a good opportunity to test the command after the Merge.

**Comment by fabioz** [comment]
Hey @OvidiuCode, not sure if you missed the:

> Note: auto-generated by `_gen_workflows.py`

(in the `**tests.yml`) or not (hopefully you didn't have to manually update all the test workflows)

If you did use it, the `_gen_workflows.py` is not committed... otherwise, please do the changes you need on `_gen_workflows.py` and then make sure you run it to update things in the test workflow files.

**Comment by OvidiuCode** [comment]
Thanks @fabioz. I missed that, that is why some were failing.

File: .github/workflows/_gen_workflows.py
Comment (cmin764) [inline .github/workflows/_gen_workflows.py]: I saw an `f""` string here but not sure what you wanted to format (as there's no variable inside). Maybe you wanted to pass in something like:
```python
"run": f"python -m pip install -r {self.devutils_requirements}",
```

File: .github/workflows/_gen_workflows.py
Comment (cmin764) [inline .github/workflows/_gen_workflows.py]: Won't this point out to: _.github/workflows/devutils/requirements.txt_ by mistake?

Maybe we want here:
```python
self.devutils_requirements = CURDIR.parent.parent / "devutils" / "requirements.txt"
```
as seen with `project_dir`.

File: .github/workflows/action_server_tests.yml
Comment (cmin764) [inline .github/workflows/action_server_tests.yml:55]: So the default current working directory is actually _.github_ while `CURDIR` points to _.github/workflows_, right?

Maybe we can also use a `ROOT = CURDIR.parent.parent` in case we use it often. (and `CURDIR` can refactor to `WORKFLOWS_DIR` as it is a little bit misleading when using it in the code and can trick you)

But your call here, if it works fine as it is, no must for a change.

**Review by cmin764 — COMMENTED** [review]
Left some comments around path resolution.

If you can temporarily trigger the workflow straight from the branch just to make sure the checks & tests are passing (at least running them so we know the workflow is good), then we're good to go!

**Review by fabioz — COMMENTED** [review]
I was just seeing the `action_server/docs/api` and I was thinking that for the action server the API isn't really helpful (as it's not a library and it's not meant to be interacted through its API, unlike the other projects in this repo).

I think we should just remove those api docs instead of updating them in this PR...

What do you think @cmin764?

**Comment by cmin764** [comment]
> I was just seeing the `action_server/docs/api` and I was thinking that for the action server the API isn't really helpful (as it's not a library and it's not meant to be interacted through its API, unlike the other projects in this repo).
> 
> I think we should just remove those api docs instead of updating them in this PR...
> 
> What do you think @cmin764?

Yes, no need to generate and store API docs for the Action Server (as it's not a library on its own), and we can skip it from docs generation during the `docs` command if we detect that there's no _docs/api_ folder present already, this way fixing the `check-all` command which will also run the `docs` in CI to ensure that there's nothing newly generated.

File: .github/workflows/_gen_workflows.py
Comment (fabioz) [inline .github/workflows/_gen_workflows.py]: I was thinking a bit about this and I don't like this approach much because if there's a single check-all which does things sequentially, you'll just get the first failure, which is not ideal.

i.e.: in the previous approach it ran the tests and then it ran the lint with a `if: always()` in the clause, so, if any of those would fail both would appear as separate failures in the ci run. With the `check-all` approach one may need one ci run to see the test failures and then another ci run to see the lint failures.

My suggestion is revert to the previous approach (ideally each separate thing that could fail which is "independent" on the others should have a separate step, so, we'd have separate steps for `tests`, `lint`, `code-format`, `typecheck` and potentially `docs`).

You could also improve the check-all to do all and report separate failures even if one fails, but I still think it's easier to see in the ci each separate failure (and one does not preclude the other -- `check-all` would be something you run locally, while the ci runs them separate).

File: log/docs/api/robocorp.log.md
Comment (fabioz) [inline log/docs/api/robocorp.log.md]: Humm, that'll change on each run -- maybe you can change the `_SentinelUseStdout` in this PR so that its `__repr__` and `__str__` returns `"_SentinelUseStdout"` so that it's possible to do the docs check and get stable results here?

Comment (OvidiuCode) [reply log/docs/api/robocorp.log.md]: sure can!

File: .github/workflows/_gen_workflows.py
Comment (cmin764) [reply .github/workflows/_gen_workflows.py]: Agree, `check-all` should be something to run locally, while having all the other sub-checks run independently in CI without a fail-fast approach, thus maximizing the CI potential of explaining to you in one go what went wrong.

File: action_server/docs/api/README.md
Comment (cmin764) [inline action_server/docs/api/README.md:1]: We can remove this also now @fabioz as agreed to strip of any generated API docs on non-libs.

**Review by cmin764 — COMMENTED** [review]
LGTM!

**Review by fabioz — APPROVED** [review]
LGTM!

**Pull request merged**
